msal_extensions/cache_lock.py

```diff
@@ -11,7 +11,7 @@ class CrossPlatLock(object):
     resource. This is specifically written to interact with a class of the same name in the .NET
     extensions library.
     """
-    def __init__(self, lockfile_path):
+    def __init__(self, lockfile_path, timeout=5):
```

Comment:
But why do we need to customize such timeout, even in our stress test? Aren't we wanna see when it would fail under NORMAL setup?

Reply:
So default timeout behaviour will basically make even 5 processes trying to access cache at the same time fail as portalocker will throw an exception. If this is configurable its useful if the use case for someone expects a load of more processes than 5 . I would say we should expose it to be set in the API but if we decide we dont want to expose it, we can guarantee only <= 5 processes to access cache roughly around the same time. Thats the limitation of the library.

Reply:
So basically this stress test would not even fail because Mac and Linux locks are not reliable after n processes. This would fail because there was a timeout.

Reply:
Thanks for this new info. That is exactly what I have been wondering, "why was there a cache failure in the first place? Shouldn't any lock mechanism raise exception when a lock isn't available, instead of silently bypassing the lock?". In that sense, that unable-to-acquire-lock exception is an expected feature, rather than a failure.

Before we make any adjustment (ideally we shouldn't have to), we want to know the behavior of our MSAL EX under default configuration. And then we go from there.

Whether we expose that timeout as a parameter, is a different topic. I doubt it would be meaningful in practice. The default 5 second timeout already makes sense.

Reply:
After a quick run on my windows machine, as the sleep interval is set as 1 sec , 5 processes is what the maximum processes can access, the lock exception is raised with 6 processes. But then again this number will depend on the sleep interval. 

I am not sure if 5 secs makes sense. It definitely does for some cases like the test case here for 5 processes, but what if there is a use case of more than 5 processes our library would fail and they would need some extra application logic to handle the timeout. In those cases if this is an expected behavior for the application instead of adding that handling logic we can have a configurable timeout instead. We would still have a default timeout of 5 secs if they don't want to configure it.

May be if we have an idea about the normal expected workload, this can be set to default to some value like 5 secs. And then we can do quick tests around how many processes are able to work under this default configuration depending on how long the lock is held.

I believe it is a configurable setting for MSAL Java and MSAL .net @bgavrilMS can confirm. Last I checked with them, java has a 6 sec default time out behavior and .net has default timeout of 60 secs but it is configurable.

Reply:
Based on the off-line discussion among us and @SomkaPe just now, let's try using the default 5 seconds timeout here (*), and:

* Have one automated test case to use `sleep_interval` as 1 second, and assert we will encounter some unable-to-acquire-lock exception when under 5+ concurrent processes. This situation is considered as a test PASS, and it would mean the locking mechanism works as expected, at least when under low load.

* Have another automated test case which lower the `sleep_interval` to very close to, or even equal to, 0 second, and to observe everything works under some perhaps ~20 processes? This test pass would mean our MSAL EX would be safe under such artificial load.

(*) If the above 2 tests pass, we do not have to expose this `timeout` parameter here.